A bit of cleanup to StopWatchingReference's code to help diagnose an issue

ToddGrun opened the pull request:
This is to help diagnose a reported issue that is difficult to track down. See https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2287350

This is primarily intended to make it a little easier to diagnose this problem if it persists. However, there is hope that this could fix the issue as it better handles if the add/remove data structures contain the same reference by handling the add processing first.

ToddGrun committed:
A bit of cleanup to StopWatchingReference's code to try and track down an issue that is being reported and difficult to track down.

ToddGrun commented on src/Workspaces/Core/Portable/Workspace/ProjectSystem/FileWatchedPortableExecutableReferenceFactory.cs:
```diff
@@ -138,23 +138,16 @@ public void StartWatchingReference(string fullFilePath)
-                if (referenceToTrack != null)
-                {
-                    // We're attempting to stop watching a file that we never started watching. This is a bug.
-                    var existingDisposalStackTrace = _previousDisposalLocations.TryGetValue(referenceToTrack, out var previousDisposalLocation);
```
This was a bit odd before in that the return value wasn't being used. Instead, this PR ensures that a referenceToTrack is passed in, and now uses the result of this TryGetValue call to determine if the reference has been seen before.

jasonmalinowski replied:
Ah yes, digging through history I see this was subject to a few rewrites and that ended up to its current state.

jasonmalinowski reviewed (changes requested):
Let's not delete this yet -- I have a work item on me right now that makes me suspect we still have a bug here.

ToddGrun commented:
> Let's not delete this yet -- I have a work item on me right now that makes me suspect we still have a bug here.

Can you clarify what deletion is concerning you? I didn't intend to delete anything, in fact, I thought I was making it more likely that we were going to be able to capture information by always passing in a reference to StopWatchingReference.

ToddGrun committed:
Reorder add/remove processing in case the same reference is in both lists.

jasonmalinowski commented:
@ToddGrun Oh I mis-read the PR and saw all the red and it looked like the disposal stuff was being removed entirely. Let me re-review.

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProjectFactory.cs:
```diff
@@ -420,21 +420,21 @@ internal void ApplyProjectUpdateState(ProjectUpdateState projectUpdateState)
     {
         Contract.ThrowIfFalse(_gate.CurrentCount == 0);
 
+        // Add file watchers for any references we are now watching. Process adds before remove in case the same reference is in both lists.
```
Should we just assert that the two lists should not contain the same thing?

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:
```diff
@@ -835,13 +839,13 @@ public static (Solution newSolution, ProjectUpdateState newProjectUpdateState) U
         // necessarily the exact same ones given to the solution itself.
```
Does this comment need updating now?

jasonmalinowski reviewed (approved):
Signing off, but I'm wondering if there's a potential alternative here, which is in that ApplyProjectUpdateState we just throw exceptions if we end up with an analyzer in both lists, and then try to chase down what's causing that in the first place. That might take a bit longer since we have to get a change into the wild with the assert, but maybe that improves long term quality?

ToddGrun committed:
1) Get rid of CWT as I don't think it's helping us diagnose this issue
2) Change ProjectUpdateState's Analyzer collections back to storing strings to reduce the churn in this PR
3) Add a bit more to comment describing why Add is done before Remove

ToddGrun committed:
Merge branch 'dotnet:main' into dev/toddgrun/ImproveDiagnosticsForFeedbackTicket2287350

ToddGrun committed:
Merge branch 'dotnet:main' into dev/toddgrun/ImproveDiagnosticsForFeedbackTicket2287350

ToddGrun committed:
Merge branch 'dotnet:main' into dev/toddgrun/ImproveDiagnosticsForFeedbackTicket2287350

ToddGrun commented:
@jasonmalinowski -- wanted to make sure this looks good as there were some changes since your last review.
